# Testing python and pandas

In [78]:
print('Python loaded correctly.')

Python loaded correctly.


In [79]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print("Packages loaded successfully.")

Packages loaded successfully.


# Loading datset

In [80]:
import pandas as pd

df = pd.read_csv('../data/raw/freMTPL2freq.csv')

# Displaying dataset information


In [81]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 678013 entries, 0 to 678012
Data columns (total 12 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   IDpol       678013 non-null  float64
 1   ClaimNb     678013 non-null  int64  
 2   Exposure    678013 non-null  float64
 3   Area        678013 non-null  str    
 4   VehPower    678013 non-null  int64  
 5   VehAge      678013 non-null  int64  
 6   DrivAge     678013 non-null  int64  
 7   BonusMalus  678013 non-null  int64  
 8   VehBrand    678013 non-null  str    
 9   VehGas      678013 non-null  str    
 10  Density     678013 non-null  int64  
 11  Region      678013 non-null  str    
dtypes: float64(2), int64(6), str(4)
memory usage: 62.1 MB


In [82]:
df.describe()

,IDpol,ClaimNb,Exposure,VehPower,VehAge,DrivAge,BonusMalus,Density
count,6.780130e+05,678013.000000,678013.000000,678013.000000,678013.000000,678013.000000,678013.000000,678013.000000
mean,2.621857e+06,0.053247,0.528750,6.454631,7.044265,45.499122,59.761502,1792.422405
std,1.641783e+06,0.240117,0.364442,2.050906,5.666232,14.137444,15.636658,3958.646564
min,1.000000e+00,0.000000,0.002732,4.000000,0.000000,18.000000,50.000000,1.000000
25%,1.157951e+06,0.000000,0.180000,5.000000,2.000000,34.000000,50.000000,92.000000
50%,2.272152e+06,0.000000,0.490000,6.000000,6.000000,44.000000,50.000000,393.000000
75%,4.046274e+06,0.000000,0.990000,7.000000,11.000000,55.000000,64.000000,1658.000000
max,6.114330e+06,16.000000,2.010000,15.000000,100.000000,100.000000,230.000000,27000.000000


In [83]:
df.head()

,IDpol,ClaimNb,Exposure,Area,VehPower,VehAge,DrivAge,BonusMalus,VehBrand,VehGas,Density,Region
0,1.0,1,0.10,D,5,0,55,50,B12,Regular,1217,R82
1,3.0,1,0.77,D,5,0,55,50,B12,Regular,1217,R82
2,5.0,1,0.75,B,6,2,52,50,B12,Diesel,54,R22
3,10.0,1,0.09,B,7,0,46,50,B12,Diesel,76,R72
4,11.0,1,0.84,B,7,0,46,50,B12,Diesel,76,R72


# Data Quality Checks


### Missing Values

In [84]:
missing_values = df.isnull().sum()
missing_values

IDpol         0
ClaimNb       0
Exposure      0
Area          0
VehPower      0
VehAge        0
DrivAge       0
BonusMalus    0
VehBrand      0
VehGas        0
Density       0
Region        0
dtype: int64

##### No missing data was able to be found within the dataset. No need for any missing data removal.

### Claim Count & Exposure Validity

#### What is the minimum value of our exposure? (to ensure exposure != 0 and is always >0)

In [85]:
exposure_minimum_value = df['Exposure'].min()
exposure_minimum_value

np.float64(0.00273224)

##### Answer: The minimum value of our exposure is 0.00273224, thus we can calculate claim frequency without worrying about dividing by zero or ending up with a negative claim frequency. 

#### What is the minimum value of our claim counts? (Ensuring no negative values)

In [86]:
claims_minimum_value = df['ClaimNb'].min()
claims_minimum_value

np.int64(0)

##### Answer: Our claim counts are all >=0, so claim counts appear clean and there are no negative values.

##### Amount of claim counts that are 0. 

In [87]:
claim_counts_zero = (df['ClaimNb'] == 0).sum()
claim_counts_zero

np.int64(643953)

##### Amount of claim counts that are 1. 

In [88]:
claim_counts_one = (df['ClaimNb'] == 1).sum()
claim_counts_one

np.int64(32178)

##### Amount of claim counts that are 2. 

In [89]:
claim_counts_two = (df['ClaimNb'] == 2).sum()
claim_counts_two

np.int64(1784)

##### Amount of claim counts that are >=3. 

In [90]:
claim_counts_three_plus = (df['ClaimNb'] >= 3).sum()
claim_counts_three_plus

np.int64(98)

##### Answer: Claim counts appear consistent with expectations. Most observations have 0 claims, with few claim counts of one, fewer of two, and very few being three or more. Additionally, since ClaimNb is a nonnegative count variable and most observations are zero, it is approporiate to treat this as claim frequency data rather than continuous data.

#### What is the maximum claim count within the dataset?

In [91]:
claim_count_max = df['ClaimNb'].max()
claim_count_max

np.int64(16)

##### Answer: Our claim count maximum is 16, which is quite high but not impossible. There is no need to exclude this data.

### Claim Count Distribution

#### What is our claim count distribution?

##### Amount of claim counts that are zero

In [92]:
claim_counts_zero = (df['ClaimNb'] == 0).sum()
claim_counts_zero

np.int64(643953)

##### Amount of claim counts that are one

In [93]:
claim_counts_one = (df['ClaimNb'] == 1).sum()
claim_counts_one

np.int64(32178)

##### Amount of claim counts that are two

In [94]:
claim_counts_two = (df['ClaimNb'] == 2).sum()
claim_counts_two

np.int64(1784)

##### Amount of claim counts that are >= to three

In [95]:
claim_counts_three_plus = (df['ClaimNb'] >=3).sum()
claim_counts_three_plus

np.int64(98)

##### Answer: Claim count distribution appears consistent with expectations. Claim counts tend to be zero, with few being one, even fewer being two, and less than one hundred being three or more. Furthermore, since ClaimNb is a nonnegative count variable and most observations are zero, it is appropriate to treat this as claim frequency/count data rather than continuous data.

### Checking categorical variable values

#### Area

In [96]:
Area_categories_amount = df['Area'].nunique()
Area_categories_amount

6

In [97]:
Area_categories_list = df['Area'].unique()
Area_categories_list

<StringArray>
['D', 'B', 'E', 'C', 'F', 'A']
Length: 6, dtype: str

In [98]:
Area_categories_count = df['Area'].value_counts()
Area_categories_count

Area
C    191880
D    151596
E    137167
A    103957
B     75459
F     17954
Name: count, dtype: int64

#### Region

In [99]:
Region_categories_amount = df['Region'].nunique()
Region_categories_amount

22

In [100]:
Region_categories_list = df['Region'].unique()
Region_categories_list

<StringArray>
['R82', 'R22', 'R72', 'R31', 'R91', 'R52', 'R93', 'R11', 'R24', 'R94', 'R83',
 'R54', 'R26', 'R53', 'R73', 'R42', 'R25', 'R21', 'R41', 'R43', 'R74', 'R23']
Length: 22, dtype: str

In [101]:
Region_categories_count = df['Region'].value_counts()
Region_categories_count

Region
R24    160601
R82     84752
R93     79315
R11     69791
R53     42122
R52     38751
R91     35805
R72     31329
R31     27285
R54     19046
R73     17141
R41     12990
R25     10893
R26     10492
R23      8784
R22      7994
R83      5287
R74      4567
R94      4516
R21      3026
R42      2200
R43      1326
Name: count, dtype: int64

#### Vehicle Brand

In [102]:
Vehicle_Brand_categories_amount = df['VehBrand'].nunique()
Vehicle_Brand_categories_amount

11

In [103]:
Vehicle_Brand_categories_list = df['VehBrand'].unique()
Vehicle_Brand_categories_list

<StringArray>
['B12', 'B6', 'B3', 'B2', 'B5', 'B10', 'B14', 'B13', 'B4', 'B1', 'B11']
Length: 11, dtype: str

In [104]:
Vehicle_Brand_categories_count = df['VehBrand'].value_counts()
Vehicle_Brand_categories_count

VehBrand
B12    166024
B1     162736
B2     159861
B3      53395
B5      34753
B6      28548
B4      25179
B10     17707
B11     13585
B13     12178
B14      4047
Name: count, dtype: int64

#### Vehicle Gas

In [105]:
Vehicle_Gas_categories_amount = df['VehGas'].nunique()
Vehicle_Gas_categories_amount

2

In [106]:
Vehicle_Gas_categories_list = df['VehGas'].unique()
Vehicle_Gas_categories_list

<StringArray>
['Regular', 'Diesel']
Length: 2, dtype: str

In [107]:
Vehicle_Gas_categories_count = df['VehGas'].value_counts()
Vehicle_Gas_categories_count

VehGas
Regular    345877
Diesel     332136
Name: count, dtype: int64

#### Answer: All categorical variables have clear, consistently labelled categories. The amount of records in each category is at least 1000, which is suitable for a basic grouped claim frequency analysis. Total exposure will also need to be checked in order to validate grouped claim frequencies. 

### Checking if extreumum and median values are reasonable

#### Driver Age

In [108]:
driver_age_maximum = df['DrivAge'].max()
driver_age_minimum = df['DrivAge'].min()
driver_age_median = df['DrivAge'].median()
driver_age_maximum, driver_age_minimum, driver_age_median

(np.int64(100), np.int64(18), np.float64(44.0))

#### Vehicle Age

In [109]:
vehicle_age_maximum = df['VehAge'].max()
vehicle_age_minimum = df['VehAge'].min()
vehicle_age_median = df['VehAge'].median()
vehicle_age_maximum, vehicle_age_minimum, vehicle_age_median

(np.int64(100), np.int64(0), np.float64(6.0))

#### Bonus Malus

In [110]:
bonus_malus_maximum = df['BonusMalus'].max()
bonus_malus_minimum = df['BonusMalus'].min()
bonus_malus_median = df['BonusMalus'].median()
bonus_malus_maximum, bonus_malus_minimum, bonus_malus_median

(np.int64(230), np.int64(50), np.float64(50.0))

#### Density

In [111]:
density_maximum = df['Density'].max()
density_minimum = df['Density'].min()
density_median = df['Density'].median()
density_maximum, density_minimum, density_median

(np.int64(27000), np.int64(1), np.float64(393.0))

#### Answer: All values appear consistent with expectations except for the maximum of vehicle ages, and bonus malus median is the same as bonus malus minimum. A vehicle which is 100 years old is possible, however, this is an outlier and so it must be investigated further. It must be determined if this is truly just an outlier, or a data error. To help minimize the impact from extremes such as this, vehicle age can be banded later for the model. As for bonus malus, I will just quickly check to see if this is consistent with the data or an error.

### Investigating extremes

#### Investigating VehAge extremes and ranges

In [112]:
vehicle_age_20orolder_count = (df['VehAge'] >= 20).sum()
vehicle_age_40orolder_count = (df['VehAge'] >= 40).sum()
vehicle_age_60orolder_count = (df['VehAge'] >= 60).sum()
vehicle_age_80orolder_count = (df['VehAge'] >= 80).sum()
vehicle_age_100_count = (df['VehAge'] == 100).sum()
vehicle_age_20orolder_count, vehicle_age_40orolder_count, vehicle_age_60orolder_count, vehicle_age_80orolder_count, vehicle_age_100_count

(np.int64(11687), np.int64(246), np.int64(80), np.int64(59), np.int64(25))

##### Answer: After analyzing, we can see that we have 11,687 cars above age 20, of which 246 are 40+, 80 are 60+, 59 are 80+, and lastly 25 are our maximum age of 100. Considering the database of 678,000 records, these numbers of vehicles as rare upper tail values are not unreasonable. I do not think this justifies removal from the dataset. To prevent poor modeling later, vehicles aged 20+ will be banded together to prevent oddities in claim frequency output.

#### Investigating Bonus Malus ranges

In [113]:
bonus_malus_50count = (df['BonusMalus'] == 50).sum()
bonus_malus_51to75count = df['BonusMalus'].between(51, 75).sum()
bonus_malus_76to100count = df['BonusMalus'].between(76, 100).sum()
bonus_malus_100overcount = (df['BonusMalus'] > 100).sum()
bonus_malus_50count, bonus_malus_51to75count, bonus_malus_76to100count, bonus_malus_100overcount

(np.int64(384156), np.int64(175012), np.int64(111051), np.int64(7794))

##### Answer: After analyzing, we can see that a malus of 50 makes up just over half of all records and so is the common/normal class for this dataset. This also explains why the median equals the minimum, so median = minimum is not odd for this dataset. A malus of 51-75 makes up around 25% of all records, and so is still common but definitely not the baseline value. A malus of 76-100 makes up around 16% of all records and so is uncommon and is clearly above the baseline BonusMalus value. A malus greater than 100 is rare, and should be treated as far above baseline BonusMalus values. At this point, we cannot draw conclusions such as a malus of 50 clearly represents good behaviour without observing claim frequency. 

#### Investigating (Population) Density extremes and median

In [114]:
density_maximum = df['Density'].max()
density_minimum = df['Density'].min()
density_median = df['Density'].median()
density_75_percentile = df['Density'].quantile(0.75)
density_25_percentile = df['Density'].quantile(0.25)
density_minimum, density_25_percentile, density_median, density_75_percentile, density_maximum

(np.int64(1),
 np.float64(92.0),
 np.float64(393.0),
 np.float64(1658.0),
 np.int64(27000))

##### A quick analysis shows that the 25th percentile is 92, the median is 393, and the 75th percentile is 1658. We can use these values to help analyze the bands of density and count how many high density urban areas we're working with. Furthermore, we can also see that our maximum value od 27,000 is not considered odd for a high density urban area, and our density of 1 is normal for people living far away from urban areas. We have no negative values as well, so the records for density are clean.

# Banding Data Together

#### Driver Age

In [115]:
## Assigning counts to age bands. 
DAge_16to24_count = df['DrivAge'].between(16, 24).sum()
DAge_25to34_count = df['DrivAge'].between(25, 34).sum()
DAge_35to44_count = df['DrivAge'].between(35, 44).sum()
DAge_45to54_count = df['DrivAge'].between(45, 54).sum()
DAge_55to64_count = df['DrivAge'].between(55, 64).sum()
DAge_65AndOver_count = (df['DrivAge'] >= 65).sum()
DAge_16to24_count, DAge_25to34_count, DAge_35to44_count, DAge_45to54_count, DAge_55to64_count, DAge_65AndOver_count

(np.int64(30198),
 np.int64(140947),
 np.int64(170578),
 np.int64(164034),
 np.int64(99094),
 np.int64(73162))

In [116]:
## Actually banding the data together
df["DriverAgeBand"] = ""
df.loc[df["DrivAge"].between(16, 24), "DriverAgeBand"] = "16-24"
df.loc[df["DrivAge"].between(25, 34), "DriverAgeBand"] = "25-34"
df.loc[df["DrivAge"].between(35, 44), "DriverAgeBand"] = "35-44"
df.loc[df["DrivAge"].between(45, 54), "DriverAgeBand"] = "45-54"
df.loc[df["DrivAge"].between(55, 64), "DriverAgeBand"] = "55-64"
df.loc[df["DrivAge"] >= 65, "DriverAgeBand"] = "65+"


In [117]:
## Testing the banding to make sure it worked. 
df["DriverAgeBand"].value_counts()

DriverAgeBand
35-44    170578
45-54    164034
25-34    140947
55-64     99094
65+       73162
16-24     30198
Name: count, dtype: int64

In [118]:
## Was anyone excluded from the banding? IF it outputs 0, then no.
(df["DriverAgeBand"] == "").sum()

np.int64(0)

##### Comment: Although these driver bands may be a bit tight, I felt as if doing more for a simple poisson model is overkill and uses up extra computational power which is unneeded for a model which already has many limitations.

#### Vehicle Age

In [119]:
## Creating vehicle age band counts
VAge_0to4_count = df['VehAge'].between(0, 4).sum()
VAge_5to9_count = df['VehAge'].between(5, 9).sum()
VAge_10to14_count = df['VehAge'].between(10, 14).sum()
VAge_15to19_count = df['VehAge'].between(15, 19).sum()
VAge_20AndOver_count = (df['VehAge'] >= 20).sum()
VAge_0to4_count, VAge_5to9_count, VAge_10to14_count, VAge_15to19_count, VAge_20AndOver_count

(np.int64(281900),
 np.int64(171936),
 np.int64(148835),
 np.int64(63655),
 np.int64(11687))

In [120]:
## Creating vehicle age bands
df["VehicleAgeBand"] = ""
df.loc[df["VehAge"].between(0, 4), "VehicleAgeBand"] = "0-4"
df.loc[df["VehAge"].between(5, 9), "VehicleAgeBand"] = "5-9"
df.loc[df["VehAge"].between(10, 14), "VehicleAgeBand"] = "10-14"
df.loc[df["VehAge"].between(15, 19), "VehicleAgeBand"] = "15-19"
df.loc[df["VehAge"] >= 20, "VehicleAgeBand"] = "20+"

In [121]:
## Testing to see if the banding worked
df["VehicleAgeBand"].value_counts()

VehicleAgeBand
0-4      281900
5-9      171936
10-14    148835
15-19     63655
20+       11687
Name: count, dtype: int64

In [122]:
## Making sure no vehicles were excluded from banding, if it outputs 0 then everything is good.
(df["VehicleAgeBand"] == "").sum()

np.int64(0)

##### Comment: Vehicle age was banded in intervals of five, and we can define a vehicle age of 0-4 as a new car, 5-9 as a used car, 10-14 as a seasoned car, 15-19 as an old car, and 20+ as a very old car. 

#### Bonus Malus

In [123]:
## Creating bonus malus band counts
bonus_malus_50count = (df['BonusMalus'] == 50).sum()
bonus_malus_51to75count = df['BonusMalus'].between(51, 75).sum()
bonus_malus_76to100count = df['BonusMalus'].between(76, 100).sum()
bonus_malus_100overcount = (df['BonusMalus'] > 100).sum()
bonus_malus_50count, bonus_malus_51to75count, bonus_malus_76to100count, bonus_malus_100overcount

(np.int64(384156), np.int64(175012), np.int64(111051), np.int64(7794))

In [124]:
## Creating bands for bonus malus
df["BonusMalusBand"] = ""
df.loc[df["BonusMalus"] == 50, "BonusMalusBand"] = "50"
df.loc[df["BonusMalus"].between(51, 75), "BonusMalusBand"] = "51-75"
df.loc[df["BonusMalus"].between(76, 100), "BonusMalusBand"] = "76-100"
df.loc[df["BonusMalus"] > 100, "BonusMalusBand"] = "101+"
       

In [125]:
## Checking to see if the bonus malus banding worked
df["BonusMalusBand"].value_counts()

BonusMalusBand
50        384156
51-75     175012
76-100    111051
101+        7794
Name: count, dtype: int64

In [126]:
## Checking to see if anyone was excluded from the bonus malus banding, if it outputs 0 then everything is good.
(df["BonusMalusBand"] == "").sum()

np.int64(0)

##### Comment: These are the same bands used to analyze bonus malus, as I felt these ranges gave satisfactory sample sizes for each band and will produce accurate enough results. 

#### Density

In [127]:
## Creating population density band counts
density_rural_count = (df['Density'] <= 100).sum()
density_low_count = df['Density'].between(101, 500).sum()
density_lowtomoderate_count = df['Density'].between(501, 1000).sum()
density_moderate_count = df['Density'].between(1001, 2500).sum()
density_moderatetohigh_count = df['Density'].between(2501, 5000).sum()
density_high_count = df['Density'].between(5001, 10000).sum()
density_veryhigh_count = df['Density'].between(10001, 15000).sum()
density_extremelyhigh_count = (df['Density'] > 15000).sum()
density_rural_count, density_low_count, density_lowtomoderate_count, density_moderate_count, density_moderatetohigh_count, density_high_count, density_veryhigh_count, density_extremelyhigh_count

(np.int64(179589),
 np.int64(191804),
 np.int64(72495),
 np.int64(92920),
 np.int64(91821),
 np.int64(31430),
 np.int64(3258),
 np.int64(14696))

In [128]:
## Creating population density bands
df["PopulationDensityBand"] = ""
df.loc[df["Density"] <= 100, "PopulationDensityBand"] = "Rural"
df.loc[df["Density"].between(101, 500), "PopulationDensityBand"] = "Low"
df.loc[df["Density"].between(501, 1000), "PopulationDensityBand"] = "Low to Moderate"
df.loc[df["Density"].between(1001, 2500), "PopulationDensityBand"] = "Moderate"
df.loc[df["Density"].between(2501, 5000), "PopulationDensityBand"] = "Moderate to High"
df.loc[df["Density"].between(5001, 10000), "PopulationDensityBand"] = "High"
df.loc[df["Density"].between(10001, 15000), "PopulationDensityBand"] = "Very High"
df.loc[df["Density"] > 15000, "PopulationDensityBand"] = "Extremely High"


In [129]:
## Ensuring that the population density banding worked
df["PopulationDensityBand"].value_counts()
## Counts agree, so banding worked fine

PopulationDensityBand
Low                 191804
Rural               179589
Moderate             92920
Moderate to High     91821
Low to Moderate      72495
High                 31430
Extremely High       14696
Very High             3258
Name: count, dtype: int64

In [130]:
## Ensuring nobody was excluded from the population density banding, if it outputs 0 then everything is good.
(df["PopulationDensityBand"] == "").sum()

np.int64(0)

##### Comment: These bands are going off how rural, suburban, urban, etc. population densities are defined, and splitting those definitions in two. Intuition says that population density of where someone lives is likely a significant risk factor, so increasing the accuracy and tightening the bands on population density is a justifiable move especially for a simple poisson distribution model.

### SQL was used to further explore exposure-adjusted claim frequency summaries the data before moving into Python modelling. In particular, the questions answered by SQL exploration can be found below.

### Importing SQLite

In [131]:
import sqlite3

conn = sqlite3.connect("../data/processed/auto_insurance_claims.db")

df.to_sql("policies", conn, if_exists="replace", index=False)

678013

### What is overall claim frequency? 

In [132]:
overall_frequency_sql = """
SELECT
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) / SUM(Exposure) AS claim_frequency
FROM policies;
"""

pd.read_sql_query(overall_frequency_sql, conn)

,total_claims,total_exposure,claim_frequency
0,36102,358499.445463,0.100703


##### This SQL query calculates our total claims, total exposure, and then calculations claim frequency directly from the policy records data we have. 

### What is claim frequency by BonusMalusBand?

In [133]:
bonus_malus_frequency_sql = """
SELECT
    BonusMalusBand,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY BonusMalusBand
ORDER BY claim_frequency DESC;
"""

pd.read_sql_query(bonus_malus_frequency_sql, conn)

,BonusMalusBand,total_claims,total_exposure,claim_frequency
0,101+,1342,3571.319003,0.375772
1,76-100,7090,44820.338260,0.158187
2,51-75,9614,84874.657298,0.113273
3,50,18056,225233.130903,0.080166


##### 

# Grouped Claim Frequency Analysis

## Overall claim frequency

In [134]:
overall_claim_frequency = df['ClaimNb'].sum() / df['Exposure'].sum()
overall_claim_frequency

np.float64(0.1007030846404707)

## Calculating grouped claim frequencies

### Importing SQL

In [135]:
import sqlite3

conn = sqlite3.connect("../data/processed/auto_insurance_claims.db")

df.to_sql("policies", conn, if_exists="replace", index=False)

678013

### Area

In [136]:
area_claimfrequency_sql = """
SELECT
    Area,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY Area
ORDER BY Area;
"""

area_claimfrequency = pd.read_sql_query(area_claimfrequency_sql, conn)
area_claimfrequency

,Area,total_claims,total_exposure,claim_frequency
0,A,5063,61969.377712,0.081702
1,B,3800,43012.323931,0.088347
2,C,9875,104449.003785,0.094544
3,D,8428,77120.191692,0.109284
4,E,7805,63819.314270,0.122298
5,F,1131,8129.234074,0.139127


### Region

In [137]:
region_claimfrequency_sql = """
SELECT
    Region,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY Region
ORDER BY Region;
"""

region_claimfrequency = pd.read_sql_query(region_claimfrequency_sql, conn)
region_claimfrequency

,Region,total_claims,total_exposure,claim_frequency
0,R11,3978,30208.641954,0.131684
1,R21,160,1205.831509,0.132689
2,R22,439,3574.410221,0.122817
3,R23,303,3177.682181,0.095353
4,R24,9204,102712.903386,0.089609
5,R25,633,6657.901142,0.095075
6,R26,511,5025.358609,0.101684
7,R31,1176,11497.974269,0.102279
8,R41,611,8113.777972,0.075304
9,R42,133,1209.115823,0.109998


### Vehicle Brand

In [138]:
vehiclebrand_claimfrequency_sql = """
SELECT
    VehBrand,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY VehBrand
ORDER BY VehBrand;
"""

vehiclebrand_claimfrequency = pd.read_sql_query(vehiclebrand_claimfrequency_sql, conn)
vehiclebrand_claimfrequency

,VehBrand,total_claims,total_exposure,claim_frequency
0,B1,8677,95351.059578,0.091001
1,B10,858,9496.014123,0.090354
2,B11,721,6887.817279,0.104678
3,B12,8859,64808.608718,0.136695
4,B13,649,6772.032216,0.095835
5,B14,166,2274.467743,0.072984
6,B2,8560,94864.057101,0.090234
7,B3,2818,28592.876331,0.098556
8,B4,1312,13775.496080,0.095242
9,B5,2020,19992.062695,0.101040


### Vehicle Gas

In [139]:
vehiclegas_claimfrequency_sql = """
SELECT
    VehGas,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY VehGas
ORDER BY VehGas;
"""

vehiclegas_claimfrequency = pd.read_sql_query(vehiclegas_claimfrequency_sql, conn)
vehiclegas_claimfrequency

,VehGas,total_claims,total_exposure,claim_frequency
0,Diesel,16648,170660.893127,0.097550
1,Regular,19454,187838.552336,0.103568


### Driver Age Band

In [140]:
driverageband_claimfrequency_sql = """
SELECT
    DriverAgeBand,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY DriverAgeBand
ORDER BY DriverAgeBand;
"""

driverageband_claimfrequency = pd.read_sql_query(driverageband_claimfrequency_sql, conn)
driverageband_claimfrequency

,DriverAgeBand,total_claims,total_exposure,claim_frequency
0,16-24,2364,12489.161085,0.189284
1,25-34,6396,64598.614437,0.099011
2,35-44,8145,87391.452517,0.093201
3,45-54,9513,89852.746482,0.105873
4,55-64,5124,56176.951112,0.091212
5,65+,4560,47990.519831,0.095019


### Vehicle Age Band

In [141]:
vehicleageband_claimfrequency_sql = """
SELECT
    VehicleAgeBand,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY VehicleAgeBand
ORDER BY VehicleAgeBand;
"""

vehicleageband_claimfrequency = pd.read_sql_query(vehicleageband_claimfrequency_sql, conn)
vehicleageband_claimfrequency

,VehicleAgeBand,total_claims,total_exposure,claim_frequency
0,0-4,15506,129017.413924,0.120185
1,10-14,7764,86098.453774,0.090176
2,15-19,2517,35981.164321,0.069953
3,20+,401,7146.975648,0.056108
4,5-9,9914,100255.437795,0.098887


### Bonus Malus Band

In [142]:
bonusmalusband_claimfrequency_sql = """
SELECT
    BonusMalusBand,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY BonusMalusBand
ORDER BY BonusMalusBand;
"""

bonusmalusband_claimfrequency = pd.read_sql_query(bonusmalusband_claimfrequency_sql, conn)
bonusmalusband_claimfrequency

,BonusMalusBand,total_claims,total_exposure,claim_frequency
0,101+,1342,3571.319003,0.375772
1,50,18056,225233.130903,0.080166
2,51-75,9614,84874.657298,0.113273
3,76-100,7090,44820.338260,0.158187


### Population Density Band

In [143]:
populationdensityband_claimfrequency_sql = """
SELECT
    PopulationDensityBand,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY PopulationDensityBand
ORDER BY PopulationDensityBand;
"""

populationdensityband_claimfrequency = pd.read_sql_query(populationdensityband_claimfrequency_sql, conn)
populationdensityband_claimfrequency

,PopulationDensityBand,total_claims,total_exposure,claim_frequency
0,Extremely High,937,6789.314567,0.138011
1,High,1816,14139.067760,0.128438
2,Low,9874,104411.194651,0.094568
3,Low to Moderate,3974,36793.095132,0.108009
4,Moderate,5243,47099.757221,0.111317
5,Moderate to High,5191,42859.710368,0.121116
6,Rural,8873,105067.386256,0.084451
7,Very High,194,1339.919507,0.144785


### Vehicle Power

In [144]:
vehiclepower_claimfrequency_sql = """
SELECT
    VehPower,
    SUM(ClaimNb) AS total_claims,
    SUM(Exposure) AS total_exposure,
    SUM(ClaimNb) * 1.0 / SUM(Exposure) AS claim_frequency
FROM policies
GROUP BY VehPower
ORDER BY VehPower;
"""

vehiclepower_claimfrequency = pd.read_sql_query(vehiclepower_claimfrequency_sql, conn)
vehiclepower_claimfrequency

,VehPower,total_claims,total_exposure,claim_frequency
0,4,5699,60073.637304,0.094867
1,5,7278,68173.348821,0.106757
2,6,8381,82524.210172,0.101558
3,7,7627,77950.456700,0.097844
4,8,1922,22684.616305,0.084727
5,9,1754,15348.213656,0.114280
6,10,1789,15375.010824,0.116358
7,11,897,8497.778452,0.105557
8,12,359,3792.038343,0.094672
9,13,153,1637.255322,0.093449


### Overall, a few, although expected, variables appear to have an association with more frequent or less frequent claims. 
### We would expect that people who live in more urban areas, with larger population densities, to have more frequent claims. Despite no formal testing, observations suggest this is true. 
### Similar to population density and area, when grouping by driver age and vehicle age (separately) it appears as though these both are associated claim frequency. Once again, despite no testing, it appears as if newer drivers tend to have more claims per policy year, and the same seems to hold true with newer vehicles. Older people and older vehicles appear to have less claims than younger. 
### Without any formal testing, the variable which appears to have the strongest association with claim frequency is bonus malus. Logically, this would make sense, as bonus malus reflects a driver's history. The association appears to be higher malus, more claims per policy year. 
### Overall, upon simply observing the data, it appears as though the variables Density, Area, Region, VehAge, DriverAge, and BonusMalus are associated with claim frequency and are worth investigating further. 

# Preliminary Poisson Frequency Model

## This section will create a preliminary poisson claim frequency model. Claim count is going to be used as the response variable, and exposure will also be included as policies are observed for different lengths of time. *This model is only a baseline frequency model and not a full pricing model*. 

## Model Setup

##### ClaimNb = response variable
##### Exposure = offset / exposure adjustment 

##### Predictors: VehicleAgeBand, DriverAgeBand, BonusMalusBand, VehBrand, VehGas, PopulationDensityBand

## Fitting the Poisson Model

#### Importing python packages

In [145]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np

#### Creating Poisson model

In [146]:
poisson_model = smf.glm(
    formula='ClaimNb ~ C(VehicleAgeBand) + C(DriverAgeBand) + C(BonusMalusBand, Treatment(reference="50")) + C(VehBrand) + C(VehGas) + C(PopulationDensityBand, Treatment(reference="Low"))',
    data=df,
    family=sm.families.Poisson(),
    offset=np.log(df["Exposure"])
).fit()

### Interpreting the Model

##### This model uses claim count as the response variable, and the log of exposure as an offset. The model is estimating claim frequency based on different policy exposure lengths. All variables are interpreted relative to their baseline categories, so coefficients should be read as relative changes to the baseline level for each category. 
##### The following variables were excluded from the model with an explanation next to them: 
##### Area and Region: Area, Region and PopulationDensityBand overlap very closely as they're all geographic location variables. This caused many large coefficients which were unstable. In order to fix the model and maintain stability PopulationDensityBand was chosen since it is a variable which is directly measurable. 
##### VehPower: In the analysis it did not show a clear association with expected claim frequency, and upon adding it to the model it did not make much, if any, difference on other variables. To prevent overcomplication and potential false associations, it has been excluded. 

#### The baselines for each category are as follows: 
###### Area: A 
###### DriverAgeBand: 16-24
###### VehicleAgeBand: 0-4
###### BonusMalusBand: 50 
###### VehBrand: B1
###### VehGas: Diesel 
###### PopulationDensityBand: Extremely High

#### Below we can view the coefficients of each categorical variable compared to our baselines.

In [147]:
model_coefficients = np.exp(poisson_model.params)
model_coefficients

Intercept                                                                   0.085280
C(VehicleAgeBand)[T.10-14]                                                  0.790362
C(VehicleAgeBand)[T.15-19]                                                  0.606928
C(VehicleAgeBand)[T.20+]                                                    0.527036
C(VehicleAgeBand)[T.5-9]                                                    0.884249
C(DriverAgeBand)[T.25-34]                                                   0.675846
C(DriverAgeBand)[T.35-44]                                                   0.860656
C(DriverAgeBand)[T.45-54]                                                   1.095165
C(DriverAgeBand)[T.55-64]                                                   0.949096
C(DriverAgeBand)[T.65+]                                                     1.034822
C(BonusMalusBand, Treatment(reference="50"))[T.101+]                        5.198369
C(BonusMalusBand, Treatment(reference="50"))[T.51-75]            

 #### Area coefficients interpretation

##### Area A is the reference category. The area values below are exponentiated coefficients, so they can be interpreted as claim frequency relativities compared with Area A, while holding the other model variables constant. 
##### Area B has a relativity of 1.043782, which reflects a minimal increase in claim frequency from our baseline of Area A, holding all other variables constant. On the other hand, Area E has a very large relativity of 3.026643. This reflects a 3x increase from A in claim frequency, holding all else constant. Other relativities are also different from baseline, with Area D's at 2.942593, F at 1.706520, and C at 1.349360.
##### Overall findings suggest that area is meaningfully associated with claim frequency.
##### One possible explanation for this association is that different geographical areas have different traffic flows. More intense traffic flows could plausibly cause more accidents, even if drivers are experienced. 

#### DriverAgeBand coefficients interpretation

In [148]:
DriverAgeBand_coefficients = np.exp(poisson_model.params).filter(like="DriverAgeBand")
DriverAgeBand_coefficients

C(DriverAgeBand)[T.25-34]    0.675846
C(DriverAgeBand)[T.35-44]    0.860656
C(DriverAgeBand)[T.45-54]    1.095165
C(DriverAgeBand)[T.55-64]    0.949096
C(DriverAgeBand)[T.65+]      1.034822
dtype: float64

##### DriverAgeBand used 16-24 as the baseline age range, thus each coefficient will be claim frequency compared to claim frequency of 16-24 year olds, *all else held constant*. 
##### Drivers aged 25-34 had a relativity of 0.675846 and drivers aged 35-44 had a relativity of 0.860656. Both of these mean that compared to drivers aged 16-24, they each have far lower claim frequency. The exact interpretation is that drivers aged 25-34 have ~67.5846% of the claim frequency that drivers aged 16-24 do. The age bands of 45-54 and 65+ have a slightly higher relativity than baseline, and the age band of 55-64 has a slightly lower relativity than baseline.
##### Overall findings suggest that driver age is meaningfully associated with claim frequency.
##### One possible explanation for this association could be that younger drivers are less experienced, so more accidents, while middle aged drivers have good reaction time and experience, while older drivers have weaker reaction times.

#### VehicleAgeBand coefficients interpretation

In [149]:
VehicleAgeBand_coefficients = np.exp(poisson_model.params).filter(like="VehicleAgeBand")
VehicleAgeBand_coefficients

C(VehicleAgeBand)[T.10-14]    0.790362
C(VehicleAgeBand)[T.15-19]    0.606928
C(VehicleAgeBand)[T.20+]      0.527036
C(VehicleAgeBand)[T.5-9]      0.884249
dtype: float64

##### VehicleAgeBand used 0-4 as the baseline, which means that *all else held constant*, each age band was compared to the claim frequency of a car aged 0-4 years old. 
##### The relativities of vehicle ages appear to decrease from baseline the older a vehicle is. 5-9 has a relativity of 0.884249, which is the smallest change from baseline. The other relativities were 0.790362 for 10-14, 0.606928 for 15-19, and 0.527036 for 20+ which was the largest change from baseline. 
##### Findings so far suggest that there is a meaningful association between vehicle age and claim frequency.
##### One possible explanation for this association is the driving and repair patterns of older cars versus newer cars, though the information in the dataset is far too limited to find the exact cause. 

#### BonusMalusBand coefficients interpretation

In [150]:
BonusMalusBand_coefficients = np.exp(poisson_model.params).filter(like="BonusMalusBand")
BonusMalusBand_coefficients

C(BonusMalusBand, Treatment(reference="50"))[T.101+]      5.198369
C(BonusMalusBand, Treatment(reference="50"))[T.51-75]     1.614284
C(BonusMalusBand, Treatment(reference="50"))[T.76-100]    2.265243
dtype: float64

##### The baseline for BonusMalus used was 50, which means *all else held constant*, each Bonus Malus range's frequency was compared to the claim frequency of 50. 
##### As for relativities, 51-75 had a relativity of 1.614284, 76-100 was 2.265243, and 101+ was 5.198369. These are large relativities in comparison to the other relativities of other variables, which suggests a very strong association with claim frequency. 
##### Findings thus far suggest that BonusMalus has a very strong association with claim frequency, with larger bonus malus values being associated with higher claim frequency. 
##### One possible explanation for this association is that bonus malus reflects a driver's history, so more risky history could be associated with higher claim frequency. 

#### VehBrand coefficients interpretation

In [151]:
VehBrand_coefficients = np.exp(poisson_model.params).filter(like="VehBrand")
VehBrand_coefficients

C(VehBrand)[T.B10]    1.029697
C(VehBrand)[T.B11]    1.124820
C(VehBrand)[T.B12]    1.240113
C(VehBrand)[T.B13]    1.050951
C(VehBrand)[T.B14]    0.838898
C(VehBrand)[T.B2]     0.988232
C(VehBrand)[T.B3]     1.004211
C(VehBrand)[T.B4]     0.959455
C(VehBrand)[T.B5]     1.061546
C(VehBrand)[T.B6]     0.948814
dtype: float64

##### The baseline for VehBrand was B1, which means *all else held constant*, each vehicle brand's claim frequency will be compared to the claim frequency of brand B1. 
##### Relativities tend to cluster around 1 for vehicle brand, however there are a few notable relativities. These notable relativities are of B11, B12, and B14 which have relativities of 1.124820, 1.240113 and 0.838898 respectively. 
##### Overall findings suggest that vehicle brand may be weakly associated with claim frequency. 

#### VehGas coefficients interpreation

In [152]:
VehGas_coefficients = np.exp(poisson_model.params).filter(like="VehGas")
VehGas_coefficients

C(VehGas)[T.Regular]    1.052343
dtype: float64

##### There are only two types of vehicle gas, that being regular and diesel. The model uses diesel as the baseline. 
##### The relativity for regular gas was 1.052343, which suggests that vehicles using regular gasoline may have slightly higher claim frequency than vehicles using diesel gasoline. 

#### PopulationDensityBand coefficients interpretation

In [153]:
population_density_coefficients = np.exp(poisson_model.params).filter(like="PopulationDensityBand")
population_density_coefficients

C(PopulationDensityBand, Treatment(reference="Low"))[T.Extremely High]      1.132079
C(PopulationDensityBand, Treatment(reference="Low"))[T.High]                1.140822
C(PopulationDensityBand, Treatment(reference="Low"))[T.Low to Moderate]     1.083128
C(PopulationDensityBand, Treatment(reference="Low"))[T.Moderate]            1.094093
C(PopulationDensityBand, Treatment(reference="Low"))[T.Moderate to High]    1.140636
C(PopulationDensityBand, Treatment(reference="Low"))[T.Rural]               0.945036
C(PopulationDensityBand, Treatment(reference="Low"))[T.Very High]           1.229126
dtype: float64

##### The population density band which was used as baseline is "Low", which means that *all else held constant*, each population density band will be compared to "Low". 
##### Relativities are mostly consistent with expectations, and are moderately higher or lower than baseline with none being extremely larger or smaller compared to baseline. Surprisingly, the largest relativity is "Very High" at 1.229126, which is larger than "Extremely High" at 1.132079. Other relativities are within about 0.1 from baseline.
##### Overall findings suggest that population density may have a moderate association with claim frequency, with "Very High" population density being associated with the highest expected claim frequency compared to baseline. 
##### One possible explanation for findings here is that a higher population density may be associated with more traffic and thus more chances for accidents to occur. A possible explanation for "Very High" density having higher relativity than "Extremely High" could be that "Extremely High" population density areas have more robust public transit and thus less vehicles the road. 

## Limitations

##### There are several limitations of this model. The first, and most trivial to mention, is that this is a preliminary claim frequency model *only*, and is not a claim severity or even a pricing model. This model predicts expected claim frequency per policy year, nothing more. 
##### The Poisson distribution (used for modelling) assumes that mean ≈ variance, which is not always true for insurance claim data. 
##### Banding variables was able to improve interpretation, however in the process it also reduced detail in the predictions. Tighter bands, or removing them altogether, could reduce the severity of this limitation, but interpretability is also important for a model.
##### Area and region, which could be two large predictors, had to be left out to avoid geographic area overlap and unstable predictions. 
##### This model shows *associations*, no *causation* has been proven with this model. Any causation is purely speculative. 
##### In the future, better data analysis, severity and pricing modelling, overdispersion checks, and validation are all good areas to improve upon. 

## Assumptions

##### 1. Claim counts follow a Poisson distribution.
##### 2. Data observations are independent of one another.
##### 3. The mean claim count is directly proportional to exposure, which is why log(Exposure) was used as an offset.
##### 4. The logarithm of expected claim frequency is linear in the predictors. 
##### 5. Mean and variance are roughly equal (refer to assumption 1)
##### 6. Selected variables capture enough of the risk differences. 
##### 7. Banded variables are reasonable representations of the original variables 

## Conclusions

##### After a very long exploratory data analysis, cleanup, and model fitting, we're able to draw several conclusions from our findings. The first, although perhaps expected conclusion, is that BonusMalus appeared to have the strongest association with expected claim frequency. For future models, a client’s claim history is a great way to predict future claims. We were also able to observe that driver age appeared to show an almost a U-shaped association with driver age. Younger drivers, as expected, tended to have the highest expected claim frequency, with middle aged drivers having the lowest. Once middle age was passed, the association started to increase once again, where older drivers tended to have a slightly higher claim frequency. Vehicle age, too, appeared to show a strong association with expected claim frequency. Newer vehicles had much higher claim frequency than older cars, and as cars got older and older claim frequency tended to decrease significantly. However, we cannot conclude whether having an older car (20+) truly leads to lower claim frequency, or whether the pattern is partly due to a smaller sample size. A variable which appeared to have a moderate association with claim frequency was population density. Higher density areas had more claims, with the highest being within the range of 10,001 - 15,000. Density bands higher than that had slightly lower, although still high claim frequency. Vehicle brand and gas type had minor associations with claim frequency, but were still useful nonetheless. 
##### Overall, this is a useful preliminary model for estimating expected claim frequency which demonstrates the ability to work with work with real data, recognize limitations and assumptions, and draw useful conclusions and improvements for the future. 